time_integration/one_step_theta.h

```diff
@@ -0,0 +1,206 @@
+
+    mutable double dt_;
+    mutable double old_time_;
+
+    mutable LinearAlgebra::distributed::Vector<Number> old_solution_;
+    mutable LinearAlgebra::distributed::Vector<Number> right_hand_side_;
+    const LevelSetProblemParameters &param_;
+
+    ConditionalOStream pcout;
+    mutable int min_iterations = std::numeric_limits<double>::infinity();
+    mutable int total_iterations = 0;
+    mutable int max_iterations = 0;
+    mutable int counter_ = 0;
+
+
+};
+
+template <typename Operator, int dim, int fe_degree>
+OneStepTheta<Operator, dim, fe_degree>::OneStepTheta(Operator &pde_operator,
+                                                     const LevelSetProblemParameters &param,
+                                                     LinearAlgebra::distributed::Vector<Number> const &) : pde_operator_(pde_operator),
+                                                                                                           param_(param),
+                                                                                                           pcout(std::cout, Utilities::MPI::this_mpi_process(MPI_COMM_WORLD) == 0),
+                                                                                                           Theta_(param.Theta_advection) {}
+
+template <typename Operator, int dim, int fe_degree>
+OneStepTheta<Operator, dim, fe_degree>::OneStepTheta(Operator &pde_operator,
+                                                     const LevelSetProblemParameters &param) : pde_operator_(pde_operator),
+                                                                                                 param_(param),
+                                                                                                 pcout(std::cout, Utilities::MPI::this_mpi_process(MPI_COMM_WORLD) == 0),
+                                                                                                 Theta_(param.Theta_IMEX) {}
+
+/* Left hand operator of the solver*/
+template <typename Operator, int dim, int fe_degree>
+void OneStepTheta<Operator, dim, fe_degree>::vmult(LinearAlgebra::distributed::Vector<Number> &dst,
+                                                   const LinearAlgebra::distributed::Vector<Number> &src) const
+{
+
+
+    dst = 0;
+    pde_operator_.apply_operator(old_time_ + dt_, dst, src);
+
+
+pde_operator_.operator_data_.data_.cell_loop(
+      &OneStepTheta<Operator,dim, fe_degree>::local_apply_inverse_mass_matrix,
+      this,
+      dst,
+      dst);
+    dst *= -1.0;
+    dst *= Theta_ * dt_;
+    dst.add(1.0, src);
+
+        
+
+}
+
+/* Creates the right hand side of the solver*/
+template <typename Operator, int dim, int fe_degree>
+void OneStepTheta<Operator, dim, fe_degree>::create_right_hand_side(bool const zero_out) const
+{
+    right_hand_side_.reinit(old_solution_);
+    if (zero_out)
+    {
+        right_hand_side_ = 0;
+    }
+
+    
+     pde_operator_.apply_operator(old_time_, right_hand_side_, old_solution_);
```

Comment:
And here you could add a `pde_operator.apply_inhomogeneous_DBC(current_time_, right_hand_side_)` function where you only perform the boundary face loop with the special treatment of the inhomogeneous DBC part.  No information on the solution vector is required for that.